interp/src/stk_env.rs

```diff
@@ -645,6 +699,147 @@ impl<K: Eq + std::hash::Hash, V: Eq> Smoosher<K, V> {
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hello", 15);
+    /// a.set("hi!", 1);
+    /// a.new_scope();
+    /// a.set("bye", 0);
+    /// let mut b = a.fork();
+    /// let mut c = a.fork_from_tail();
+    /// a.set("hi!", 3);
+    /// b.set("bye", 2);
+    /// c.set("privet", 11);
+    /// let mut lst = Vec::new();
+    /// lst.push(b);
+    /// lst.push(c);
+    /// let d = Smoosher::merge_many(a, lst);
+    /// assert_eq!(*d.get(&"privet").unwrap(), 11);
+    /// assert_eq!(*d.get(&"hi!").unwrap(), 3);
+    /// assert_eq!(*d.get(&"bye").unwrap(), 2);
+    /// assert_eq!(*d.get(&"hello").unwrap(), 15);
+    /// ```
+    /// will panic:
+    /// ```should_panic
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hello", 15);
+    /// a.set("hi!", 1);
+    /// a.new_scope();
+    /// a.set("bye", 0);
+    /// let mut b = a.fork();
+    /// a.set("hi!", 3);
+    /// let mut c = a.fork();
+    /// b.set("bye", 2);
+    /// c.set("privet", 11);
+    /// let mut lst = Vec::new();
+    /// lst.push(b);
+    /// lst.push(c);
+    /// let d = Smoosher::merge_many(a, lst); //a and b has a different fork point
+    //from a and c
+    /// ```
+    pub fn merge_many(self, other: Vec<Self>) -> Self {
+        //initialize all needed variables
+        let mut a = self;
+        //needed to check for common fork point for all smooshers
+        let mut dp_first = 0;
+        let mut smooshed = Vec::new();
+
+        //iterate over all the smooshers and check for common fork point for all
+        //of them as well as smoosh them all to one level above the common fork.
+        for sm in other {
+            if let Some((depth_a, depth_b)) =
+                Smoosher::shared_fork_point(&a, &sm)
+            {
+                if dp_first == 0 {
+                    dp_first = depth_a;
+                    smooshed.push(sm.smoosh(depth_b - 1));
+                } else {
+                    if dp_first != depth_a {
+                        panic!(
+                            "The common fork differs for one or more smooshers"
+                        )
+                    } else {
+                        smooshed.push(sm.smoosh(depth_b - 1));
+                    }
+                }
+            } else {
+                panic!("No common fork for a pair of smooshers")
+            }
+        }
+
+        a = a.smoosh(dp_first - 1);
+
+        let mut a_head = a.head;
+
+        //iterate over every smooshed smoosher and put all of their values in
+        //the head of the first smoosher.
+        for sm in smooshed {
+            for (k, v) in sm.head {
+                if a_head.insert(k, v).is_some() {
+                    panic!("arguments of merge are not disjoint");
+                }
+            }
+            std::mem::drop(sm.tail);
+        }
+
+        let (a_new_head, a_new_tail) = a.tail.split();
+        if let Some(a_new_head) = a_new_head {
```

Comment:
These two lines can be merged into a single pattern match

Reply:
Changed in current and previous code, thank you